src/supermarket/app/models/cookbook_upload/archive.rb

```diff
@@ -78,26 +81,41 @@ def read(path)
     # @raise [NoPath] if the source has no path
     # @raise [Error] if the source is not a compatible archive
     #
-    # @yieldparam [::Gem::Package::TarReader::Entry] entry
+    # @yieldparam [::Archive::Entry] entry
+    # @yieldparam [::String] contents
     #
     # @example
     #   archive = CookbookUpload::Archive.new(tarball)
     #   archive.each do |entry|
     #     puts "#{entry.full_name} has the following content:\n#{entry.read}"
     #   end
     #
-    def each
+    def each_entry
       raise NoPath unless @source.respond_to?(:path)
+      raise NotGzipped unless gzipped?
 
       begin
-        Zlib::GzipReader.open(@source.path) do |gzip|
-          Gem::Package::TarReader.new(gzip) do |tar|
-            tar.each { |entry| yield entry }
+        ::Archive::Reader.open_filename(@source.path) do |tar|
+          tar.each_entry_with_data do |entry, data|
+            contents = data.is_a?(String) ? data : ''
+            yield entry, contents
           end
         end
-      rescue Zlib::GzipFile::Error => e
-        raise Error, e.message
+      rescue ::Archive::Error => e
+        case e.message
+        when /Damaged tar archive/
+          raise CorruptTarball, e.message
+        end
       end
     end
+
+    #
+    # Determines whether a file at a path is gzipped or not
+    #
+    # @return true/false
+    #
+    def gzipped?
+      FileMagic.mime.file(@source.path).match? %r{application\/x?-?gzip;}
```

Comment:
What is an example of the mime string that will be returned? I'm not familiar with FileMagic but just trying to ensure this regex will match

Reply:
I saw `"application/x-gzip; charset=binary"` on macOS and  then `"application/gzip; charset=binary"` on Linux in Travis. (The reason for several of the force pushes.) I used [Rubular to tinker with the regex](https://rubular.com/r/bNfFNbNYqqFof4).

Reply:
Added a code comment with basically those words.